hail/python/hailtop/aiotools/fs.py

```diff
@@ -673,11 +682,13 @@ async def copy(self, sema: asyncio.Semaphore, source_report: SourceReport, retur
                 return_exceptions=True)
 
             assert self.pending == 0
-            assert (self.src_is_file is None) == self.src.endswith('/')
-            assert self.src_is_dir is not None
 
-            if (self.src_is_file is False or self.src.endswith('/')) and not self.src_is_dir:
-                raise FileNotFoundError(self.src)
+            if not self.encountered_error:
```

Comment:
Then you can put the check for errors above these checks, and get rid of encountered_error.